code/__HELPERS/traits.dm

```diff
@@ -2,89 +2,116 @@
 #define SIGNAL_REMOVETRAIT(trait_ref) "removetrait [trait_ref]"
 
 // trait accessor defines
+
+/**
+ * Adds a status trait to the target datum.
+ *
+ * Arguments: (All Required)
+ * * target - The datum to add the trait to.
+ * * trait - The trait which is being added.
+ * * source - The source of the trait which is being added.
+ */
 #define ADD_TRAIT(target, trait, source) \
 	do { \
-		var/list/_L; \
-		if (!target.status_traits) { \
-			target.status_traits = list(); \
-			_L = target.status_traits; \
-			_L[trait] = list(source); \
-			SEND_SIGNAL(target, SIGNAL_ADDTRAIT(trait), trait); \
+		LAZYINITLIST(target.status_traits); \
+\
+		if(!target.status_traits[trait]) { \
+			target.status_traits[trait] = list(source); \
 		} else { \
-			_L = target.status_traits; \
-			if (_L[trait]) { \
-				_L[trait] |= list(source); \
-			} else { \
-				_L[trait] = list(source); \
-				SEND_SIGNAL(target, SIGNAL_ADDTRAIT(trait), trait); \
-			} \
+			target.status_traits[trait] |= list(source); \
 		} \
+\
+		SEND_SIGNAL(target, SIGNAL_ADDTRAIT(trait), trait); \
 	} while (0)
+
+/**
+ * Removes a status trait from a target datum.
+ *
+ * `ROUNDSTART_TRAIT` traits can't be removed without being specified in `sources`.
+ * Arguments:
+ * * target - The datum to remove the trait from.
+ * * trait - The trait which is being removed.
+ * * sources - If specified, only remove the trait if it is from this source. (Lists Supported)
+ */
 #define REMOVE_TRAIT(target, trait, sources) \
 	do { \
-		var/list/_L = target.status_traits; \
-		var/list/_S; \
-		if (sources && !islist(sources)) { \
-			_S = list(sources); \
-		} else { \
-			_S = sources\
-		}; \
-		if (_L && _L[trait]) { \
-			for (var/_T in _L[trait]) { \
-				if ((!_S && (_T != ROUNDSTART_TRAIT)) || (_T in _S)) { \
-					_L[trait] -= _T \
+		if(target.status_traits && target.status_traits[trait]) { \
+			var/list/TARGET_TRAITS = target.status_traits; \
```

Comment:
It seems you can remove this variable and just stick with using `target.status_traits`. Was there any reason you decided not to? Same goes for the two other `REMOVE_` defines below.

Reply:
In this case I believe it was to try and help with readability.
```dm
for(var/TRAIT in TARGET_TRAITS) { \
	TARGET_TRAITS[TRAIT] &= SOURCES; \
	if(!length(TARGET_TRAITS[TRAIT])) { \
		TARGET_TRAITS -= TRAIT; \
		SEND_SIGNAL(target, SIGNAL_REMOVETRAIT(TRAIT), TRAIT); \
```
vs
```dm
for(var/TRAIT in target.status_traits) { \
	target.status_traits[TRAIT] &= SOURCES; \
	if(!length(target.status_traits[TRAIT])) { \
		target.status_traits -= TRAIT; \
		SEND_SIGNAL(target, SIGNAL_REMOVETRAIT(TRAIT), TRAIT); \
```
Although looking at it now it might actually be better the second way...